modules/mod-deep-learning/DeepModelManager.cpp

```diff
@@ -0,0 +1,554 @@
+}
+
+std::string DeepModelManager::GetMoreInfoURL(ModelCardHolder card) const
+{
+   return "https://huggingface.co/" + card->GetRepoID();
+}
+
+bool DeepModelManager::IsInstalled(ModelCardHolder card) const
+{
+   FilePath repoDir = wxFileName(card->local_path()).GetFullPath();
+
+   wxFileName modelPath = wxFileName(repoDir, "model.pt");
+   wxFileName metadataPath = wxFileName(repoDir, "metadata.json");
+
+   return (modelPath.FileExists() && metadataPath.FileExists());
+}
+
+void DeepModelManager::Install(ModelCardHolder card, ProgressCallback onProgress, CompletionHandler onCompleted)
+{ 
+   if (IsInstalled(card))
+      return ;
+
+   ProgressCallback progressHandler(
+   [this, card, handler {std::move(onProgress)}](int64_t current, int64_t expected)
+   {
+      // if the install has been cancelled, bail
+      // if we don't bail early, then calling the handler will 
+      // segfault.
+      if (!this->IsInstalling(card))
+         return;
+
+      return handler(current, expected);
+   }
+   );
+
+   // set up the install handler
+   CompletionHandler installHandler(
+   [this, card, handler {std::move(onCompleted)} ](int httpCode, std::string body)
+   {
+      if (!this->IsInstalling(card))
+      {
+         Uninstall(card);
+         return;
+      }
+
+      if (!(httpCode == 200) && 
+          !(httpCode == 302))
+         Uninstall(card);
+          
+      
+      // let the caller handle this
+      handler(httpCode, body);
+
+      // get rid of the cached response
+      this->mResponseMap.erase(card->GetRepoID());
+   });
+
+   // download the model
+   try 
+   {
+      // save the metadata
+      wxLogDebug("saving modelcard for %s \n", card->GetRepoID());
+      card->SerializeToFile(
+         audacity::ToUTF8(wxFileName(GetRepoDir(card), "metadata.json").GetFullPath())
+      );
+
+      wxLogDebug("downloading model for %s \n",card->GetRepoID());
+
+      network_manager::ResponsePtr response = DownloadModel(card, progressHandler, installHandler);
+
+      // add response to a temporary map (in case of cancellation)
+      mResponseMap[card->GetRepoID()] = response;
+   }
+   catch (const char *msg)
+   {
+      wxLogError(msg);
+      return;
+   }
+}
+
+void DeepModelManager::Uninstall(ModelCardHolder card)
+{
+   bool success = true;
+   success &= wxRemoveFile(wxFileName(GetRepoDir(card), "model.pt").GetFullPath());
+   success &= wxRemoveFile(wxFileName(GetRepoDir(card), "metadata.json").GetFullPath());
+
+   if (!success)
+      throw ModelManagerException(XO("An error occurred while uninstalling the model."));
+   
+   wxRmDir(wxFileName(GetRepoDir(card)).GetFullPath());
+   
+}
+
+bool DeepModelManager::IsInstalling(ModelCardHolder card) const
+{
+   return mResponseMap.find(card->GetRepoID()) != mResponseMap.end();
+}
+
+void DeepModelManager::CancelInstall(ModelCardHolder card)
+{
+   if (!IsInstalling(card))
+    {
+      // should never really reach here (can't cancel an install that's not ongoing)
```

Comment:
I'm not very sure of this assumption, even after all is serialized in the main thread.  It's harmless if the assumption is wrong.  I wouldn't make an assertion.